### PR TITLE
add "working with bitbake" section

### PR DESCRIPTION
**Pull request by beroset**
We need a quick start guide to understanding and using bitbake in the context of AsteroidOS beyond the existing "RTFM" advice

### CODE REVIEW

**Comment by FlorentRevest** [comment]
Did you forget to add `pages/wiki/bitbake.md` to the commit ?

**Comment by beroset** [comment]
Yes, I did.  I'll mark this as draft until that's done.

File: pages/wiki/bitbake.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+Two fundamental concepts within Yocto and bitbake are the *layer* and the *recipe*.  Layers are the way that, for example, the specific watch hardware can be accomodated and then the asteroid layer can be built on top without changes.  A recipe provides instructions for creating some specific unit of software (a program or library, for instance) just as in cooking, a recipe provides instructions for how to prepare a specific food.  
+
+We can start by examining the layers.  
+
+```
+bitbake-layers show-layers
+NOTE: Starting bitbake server...
+layer                 path                                      priority
+==========================================================================
```
Comment (MagneFire) [inline pages/wiki/bitbake.md:35]: Awesome! This shows that we might have an issue with layer priorities?
@FlorentRevest do you know if this is true?

**Review by MagneFire — COMMENTED** [review]
Great work! I did not know that the `devtool` is so useful for development thanks for this!

Only have some minor comments.
Maybe we should also rename the page to `Working with OpenEmbedded` as it references that page a few times and this page is a logical extension on that?

It looks like there are double spaces here and there, is this intentional?
Additionally, `we` is used very frequently. While not particularly bad, we might want to reduce it a bit?

File: pages/wiki/documentation.md [hunk 2]
```diff
@@ -32,6 +32,7 @@ layout: documentation
   <li><a href="{{rel 'wiki/ble-profiles'}}">Bluetooth Low Energy Profiles</a></li>
   <li><a href="{{rel 'wiki/conferences'}}">Conferences</a></li>
   <li><a href="{{rel 'wiki/porting-guide'}}">Porting Guide</a></li>
+  <li><a href="{{rel 'wiki/bitbake'}}">Working with bitbake</a></li>
```
Comment (FlorentRevest) [inline pages/wiki/documentation.md:35]: I haven't read the content yet because I understand you still need to address MagneFire's comments and you have more stuff you wanted to do but just a general note, this feels like this will have a lot of duplicated content with the existing OpenEmbedded page just below. Have you considered consolidating the two pages ? I feel that duplicating the information will only lead to fragmentation and a well structured document should be able to start with a high level description suiting everyone followed by a deeper dive suiting those who are interested in it deeper.

File: pages/wiki/bitbake.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+```
+rm -rf ~/.cache/asteroid-flashlight
+```
+
+This will force the code to be turned into bytecode the next time the application runs, and indeed we can see that now our flashlight is a lovely blue color.
+
+## Sharing with others
+
+The description above is suitable for a single developer working with a single watch, but if we want to share the app, perhaps for testing, we need a better way to do this.  Once the code changes have been made as described above, w can exit the devshell and run this command:
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/bitbake.md:124]: You're right.  I've completely rewritten most of it.

File: pages/wiki/bitbake.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+``` 
+color: flashOn ? "lightblue" : "#66444444"
+```
+
+With our change to the sourcecode made in `src/main.qml`, we save the file and we are ready to rebuild the app.  Because we know that this is a CMake project (because of the existence of a CMakeLists.txt file in the source directory) we could navigate one directory up from the `git` directory we were in and run `cmake --build build` to rebuild the project in the `build` directory, but what if it's an `Autotools` project or uses `QMake`?  Fortunately, there is an easy answer.  One way to do it would be to do `cd ../temp/` to move to the temp directory that is a peer of the `git` directory the `run.do_compile` script.  This runs exactly the tasks that `bitbake` would have run, and so it works for many different kinds of builds.  We can then do `run.do_install` from the same directory.  If we observe the output carefully, we can see that it created a number of language translation files and a `.desktop` file and then finally the application itself.  The last few lines look like this:
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/bitbake.md:99]: I made what I think is an even better change and only describe how to use `devtool`.  There is therefore much less explanation of directory trees than was formerly necessary.

File: pages/wiki/bitbake.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+This will automatically install all of the install targets from the `hello` project onto the watch.  If you check the launcher and scroll to the end of your list of apps, you should now see a "HELLO WORLD APP" at the end.  Running it will show "Hello World!" on the screen.
+
+Now we can also remove it from the watch just as simply:
+
+```
+devtool undeploy-target hello root@192.168.2.15
+```
+
+This was a very brief introduction into using Yocto and bitbake in the context of AsteroidOS development.  For much more detail, refer to the Yocto manuals.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/bitbake.md:232]: Done by combining this page with the existing OpenEmbedded page (per kido's suggestion).

File: pages/wiki/bitbake.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+---
+
+This will guide you step-by-step from the process of wanting to make a change to actually making that change and getting it running on your watch.
+
+Let's say that you want to make a simple change to the flashlight app on the watch.  Your favorite color is blue, so you want to change the flashlight screen's color to a blue color.  
+
+Since you have already run `bitbake asteroid-image` you know that there are over 5000 tasks to create the image file that can be flashed to a watch.  How do you figure out which of those 5000 tasks builds the flashlight application?
+
+Two fundamental concepts within Yocto and bitbake are the *layer* and the *recipe*.  Layers are the way that, for example, the specific watch hardware can be accomodated and then the asteroid layer can be built on top without changes.  A recipe provides instructions for creating some specific unit of software (a program or library, for instance) just as in cooking, a recipe provides instructions for how to prepare a specific food.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/bitbake.md:27]: You're right.  I've removed it, since the existing OpenEmbedded page already describes these concepts.

File: pages/wiki/documentation.md (hunk 2 above)
Comment (beroset) [reply pages/wiki/documentation.md:36]: You're right.  I've combined the pages and it's much nicer now, without all of the duplication.

File: pages/wiki/bitbake.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+rm -rf ~/.cache/asteroid-flashlight
+```
+
+# Adding a New Application
+
+---
+
+Let's say you have decided to create a new application.  As described in [Creating an Asteroid app]({{rel 'wiki/creating-an-asteroid-app'}}) page, the [Asteroid-helloworld](https://github.com/AsteroidOS/asteroid-helloworld) which can be used as the basis for your own new application can be cloned and then modified.  So let's say you've done that and now you'd like to build the app, deploy it and test it.  You can, of course, follow the instructions listed on that page, but here we'll show how to do so by working within Yocto and bitbake.
+
+As mentioned above, Yocto and bitbake use *layers* and the *recipes*.  The simplest way to proceed is to use the `prepare-build.sh` file.  If you were planning, for instance on trying your app on a `sturgeon` model (a Huawei watch), you'd navigate to the directory in which the `prepare-build.sh` file exists and run:
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/bitbake.md]: Right!

**Comment by beroset** [comment]
Thanks for all of the great suggestions.  I think this is now much better as a direct result.  Thanks!

File: pages/wiki/openembedded.md [hunk 8]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+If you make a change to AsteroidOS software that you think might be useful to others, such as a bug fix or feature enhancement, there are several ways to share it, depending on what exactly is being changed.   
+
+ - creating a recipe patch file
+ - creating a pull request for an AsteroidOS project
+ - creating an installable package (`.ipk`)
+
+## Creating a recipe patch file
+
+Many of the software packages that make up an AsteroidOS image are not unique to AsteroidOS.  They may be general purpose tools such as `systemctl` or watch-specific software from other open source repositories such as `mcetool`.  For these, there may be a need to introduce a patch just for AsteroidOS.  To do so, first, within the appropriate `workspace` directory where you made changes, use `git commit` to commit the changes to the local repository.  Once this is done, automatically create the patch.  For our example, we could do that with this command:
```
Comment (MagneFire) [inline pages/wiki/openembedded.md:199]: One also needs to stage the files using `git add`.
It might be better to remove the `git` related parts and just mention that someone needs to commit the changes using git and link to a good git tutorial if need be (@FlorentRevest had one in the past, but I'm not sure where to find it anymore).

**Review by MagneFire — COMMENTED** [review]
Another excellent PR.

In this one I would also like to add that there are a few `double spaces` here, instead of mentioning them individually.

File: pages/wiki/openembedded.md [hunk 9]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
-<p>A recipe describe how to fetch, patch, configure, compile, install, package (in .rpm, .deb or .ipk) and test a piece of software. Bitbake handles all those operations and dependencies between them and between recipes. In the end it can generate images or SDK for multiple targets.</p>
-<p>Contributing to the AsteroidOS’s OE architecture mostly consists in maintaining the <a href="https://github.com/AsteroidOS/meta-asteroid">meta-asteroid</a> and meta-*-hybris repositories which contain the recipes related to Asteroid and the watches BSPs.</p>
-<p>In depth info can be found in <a href="https://docs.yoctoproject.org/index.html">the Yocto Project's manual</a>.</p>
+# OpenEmbedded
+
+[OpenEmbedded](http://www.openembedded.org/wiki/Main_Page) (OE) is the build system used by AsteroidOS. It provides a way to easily maintain embedded Linux distributions. OE is also used by project such as WebOS-Ports, SHR or Ångström. OpenEmbedded is supported by the [Yocto project](https://www.yoctoproject.org/) from the Linux Foundation and provides a rock-solid basis for AsteroidOS.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:8]: That was from the original, but you're right.  Fixed.

File: pages/wiki/openembedded.md [hunk 10]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+OpenEmbedded uses the bitbake tool which is developed in python to generate images (like asteroid-image) from packages (like asteroid-launcher-dev) built from recipes (ex: asteroid-launcher_git.bb) provided in layers (ex: meta-asteroid).
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:10]: These are all unchanged from the original, but I've added a hyperlink for `bitbake`.

File: pages/wiki/openembedded.md [hunk 11]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+Layers are git repositories containing a bunch of related recipes, example of layers include board support packages (BSPs) like [meta-dory-hybris](https://github.com/AsteroidOS/meta-smartwatch/tree/master/meta-dory) or meta-rockchip which describe a way to support a new machine, but also UI layers like meta-xfce or meta-gnome which describe the building process of graphic components.
+
+A recipe describes how to fetch, patch, configure, compile, install, package (in .rpm, .deb or .ipk) and test a piece of software. Bitbake handles all those operations and dependencies between them and between recipes. In the end it can generate images or SDK for multiple targets.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:14]: Also from the original, but now changed.

File: pages/wiki/openembedded.md [hunk 12]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+Contributing to the AsteroidOS’s OE architecture mostly consists in maintaining the [meta-asteroid](https://github.com/AsteroidOS/meta-asteroid) and [meta-smartwatch](https://github.com/AsteroidOS/meta-smartwatch) repositories which contain the recipes related to Asteroid and the watches' BSPs.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:16]: It's weird, but in the English language, plural possessive nouns are formed by adding a single apostrophe after the `s` as is done here.  For Asteroid vs. AsteroidOS, I've changed it here even though 1) this was in the original and 2) the page about creating an app is already titled "Creating an Asteroid app".

File: pages/wiki/openembedded.md [hunk 13]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
 <img src="{{assets}}/img/openembedded.png" style="width:100%"/>
+
+
+# A Quick Tutorial
+
+---
+
+
+The rest of this description assumes that you have already read and followed the [Building AsteroidOS]({{rel 'wiki/building-asteroidos'}}) instructions, and that you have built AsteroidOS for at least one platform (a particular watch or the emulator).
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:27]: Knowing how to use Linux to help develop a Linux distribution seems reflexively obvious to me.  Also, aren't those who would undertake to alter or enhance AsteroidOS are reasonably assumed to know how to use it and the web site?

File: pages/wiki/openembedded.md [hunk 14]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+meta-smelt            /home/developer/asteroid/src/meta-smartwatch/meta-smelt  7
+meta-sparrow          /home/developer/asteroid/src/meta-smartwatch/meta-sparrow  7
+meta-sprat            /home/developer/asteroid/src/meta-smartwatch/meta-sprat  7
+meta-sturgeon         /home/developer/asteroid/src/meta-smartwatch/meta-sturgeon  7
+meta-swift            /home/developer/asteroid/src/meta-smartwatch/meta-swift  8
+meta-tetra            /home/developer/asteroid/src/meta-smartwatch/meta-tetra  7
+meta-wren             /home/developer/asteroid/src/meta-smartwatch/meta-wren  7
+```
+
+By convention, Yocto layers all begin with `meta`.  There are layers for individual pieces such as Qt, networking, Python, and filesystems.  These are followed by the layers for each of the watch codenames.  If you are looking for an Asteroid-specific application, you can probably guess that it's in the `meta-asteroid` layer, but you can also ask the computer to search for you.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:80]: Done.

File: pages/wiki/openembedded.md [hunk 15]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+```
+devtool undeploy-target asteroid-flashlight root@192.168.2.15
+```
+
+# Adding a New Application
+
+---
+
+Let's say you have decided to create a new application.  As described in [Creating an Asteroid app]({{rel 'wiki/creating-an-asteroid-app'}}) page, the [Asteroid-helloworld](https://github.com/AsteroidOS/asteroid-helloworld) which can be used as the basis for your own new application can be cloned and then modified.  So let's say you've done that and now you'd like to build the app, deploy it and test it.  You can, of course, follow the instructions listed on that page, but here we'll show how to do so by working within Yocto and bitbake.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:130]: That's the existing title since 2017.  I was trying to minimize the size of this PR.

File: pages/wiki/openembedded.md [hunk 16]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+
+# A Quick Tutorial
+
+---
+
+
+The rest of this description assumes that you have already read and followed the [Building AsteroidOS]({{rel 'wiki/building-asteroidos'}}) instructions, and that you have built AsteroidOS for at least one platform (a particular watch or the emulator).
+
+This guide is for those who wish to go deeper and would like to alter or augment the AsteroidOS software.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:29]: I had them in that order originally, but "Sharing your changes" applies equally to both "Changing an Existing Application" and "Adding a New Application" so it seemed to make more sense to me to put it at the end.

File: pages/wiki/openembedded.md [hunk 17]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+devtool undeploy-target asteroid-flashlight root@192.168.2.15
+```
+
+# Adding a New Application
+
+---
+
+Let's say you have decided to create a new application.  As described in [Creating an Asteroid app]({{rel 'wiki/creating-an-asteroid-app'}}) page, the [Asteroid-helloworld](https://github.com/AsteroidOS/asteroid-helloworld) which can be used as the basis for your own new application can be cloned and then modified.  So let's say you've done that and now you'd like to build the app, deploy it and test it.  You can, of course, follow the instructions listed on that page, but here we'll show how to do so by working within Yocto and bitbake.
+
+As mentioned above, Yocto and bitbake use *layers* and the *recipes*.  The simplest way to proceed is to use the `prepare-build.sh` file.  If you were planning, for instance on trying your app on a `sturgeon` model (a Huawei Watch), you'd navigate to the directory in which the `prepare-build.sh` file exists and run:
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:132]: You're right.  I've eliminated most of this paragraph.

File: pages/wiki/openembedded.md [hunk 18]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+INFO: Recipe /home/developer/asteroid/build/workspace/recipes/hello/hello_git.bb has been automatically created; further editing may be required to make it fully functional
+```
+
+So the tool has created a recipe based on just the name we have given and the directory containing the source.  Since it seems unsure about the license and suggests that further editing might be required, let's take it's advice and edit the recipe.  The simplest way to do that is again to use `devtool`:
+
+```
+devtool edit-recipe hello
+```
+
+This will open the recipe in an editor.  It says that the license is unknown, so we change the line from `LICENSE = "Unknown"` to `LICENSE = "GPL-3.0-only"`.  Not that the `SRC_URI` is set to point to a git repository.  This is because `devtool` is smart enough to figure out that even though you gave it a local directory, it discovered that the source was actually git and made the appropriate changes.  We can now attempt to build.
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:160]: Both of these are likely to change in the future.  For this level of detail, I made a conscious decision to let the user discover the answer via a search engine rather than tying this site to particular sources like these.  This is, after all, only one of a hundred different kinds of such problems a user might encounter at this step and I'd rather not try to create a comprehensive "how to develop" page, but just concentrate on the items specific to `AsteroidOS` or this particular example.

File: pages/wiki/openembedded.md (hunk 8 above)
Comment (beroset) [reply pages/wiki/openembedded.md:199]: I'm assuming developers know how to develop, but it's not unreasonable to provide a link to some help.  Can you suggest a link?

**Comment by beroset** [comment]
> Another excellent PR.
> 
> In this one I would also like to add that there are a few `double spaces` here, instead of mentioning them individually.

Double spaces after a period are an artifact of the way I learned to type fifty years ago and not a habit I'm actively trying to alter, although I understand the conventions have (relatively) recently changed.  However, for this it truly doesn't matter since both single and double spaces are rendered *exactly* the same way on the web site.

File: pages/wiki/openembedded.md [hunk 19]
```diff
@@ -3,10 +3,245 @@ title: OpenEmbedded
...
+
+As mentioned above, Yocto and bitbake use *layers* and the *recipes*.  The simplest way to proceed is to use the `prepare-build.sh` file.  If you were planning, for instance on trying your app on a `sturgeon` model (a Huawei Watch), you'd navigate to the directory in which the `prepare-build.sh` file exists and run:
+
+```
+source ./prepare-build.sh sturgeon
+```
+
+This will set the necessary environment variables and place you in the `build` directory.  For these instructions, it is assumed that you have already built AsteroidOS for `sturgeon` (or your preferred test platform).  Let's say you have already cloned the `asteroid-helloworld` into a directory named `~/tools/AsteroidOS/asteroid-helloworld/`.  You can very easily create a recipe for it with `devtool`:
```
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md]: Right -- done.

File: pages/wiki/openembedded.md (hunk 8 above)
Comment (beroset) [reply pages/wiki/openembedded.md:193]: I added a link to an atlassian git tutorial.

File: pages/wiki/openembedded.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:29]: > > Another excellent PR.
> > In this one I would also like to add that there are a few `double spaces` here, instead of mentioning them individually.
> 
> Double spaces after a period are an artifact of the way I learned to type fifty years ago and not a habit I'm actively trying to alter, although I understand the conventions have (relatively) recently changed. However, for this it truly doesn't matter since both single and double spaces are rendered _exactly_ the same way on the web site.

I've changed to all single spaces per convention.  My fingers still type two spaces, but I search and replace pretty well.

**Comment by beroset** [comment]
> > Another excellent PR.
> > In this one I would also like to add that there are a few `double spaces` here, instead of mentioning them individually.
> 
> Double spaces after a period are an artifact of the way I learned to type fifty years ago and not a habit I'm actively trying to alter, although I understand the conventions have (relatively) recently changed. However, for this it truly doesn't matter since both single and double spaces are rendered _exactly_ the same way on the web site.

I did a search and replace to make them all single spaces.  My fingers still type double spaces, but I can search and replace!

File: pages/wiki/openembedded.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (MagneFire) [reply pages/wiki/openembedded.md]: Sounds good :+1:

File: pages/wiki/openembedded.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (MagneFire) [reply pages/wiki/openembedded.md:29]: I totally agree on the current structure.
This was more about adding a short paragraph that introduces the next sections. As a way to guide the reader.
It could be a simple list as well:

This tutorial will introduce you to the following topics:
- *Changing an existing application*: You don't like how something behaves or want to fix an issue in an existing application? This will introduce you to the tooling used to allow you to make these changes.
- *Adding a new application*: Use the tooling to create a new application recipe and test it on a watch.
- *Sharing your changes*: You've made changes on your local build of AsteroidOS. How do you share these changes? This will be explained in this part of the tutorial.

[the remark this answers is not in the document]
Comment (beroset) [reply pages/wiki/openembedded.md:29]: I like this idea and have added your text to the page, verbatim.  Thanks!

File: pages/wiki/openembedded.md [hunk 20]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
-<p>Contributing to the AsteroidOS’s OE architecture mostly consists in maintaining the <a href="https://github.com/AsteroidOS/meta-asteroid">meta-asteroid</a> and meta-*-hybris repositories which contain the recipes related to Asteroid and the watches BSPs.</p>
-<p>In depth info can be found in <a href="https://docs.yoctoproject.org/index.html">the Yocto Project's manual</a>.</p>
+# OpenEmbedded
+
+[OpenEmbedded](http://www.openembedded.org/wiki/Main_Page) (OE) is the build system used by AsteroidOS. It provides a way to easily maintain embedded Linux distributions. OE is also used by projects such as WebOS-Ports, SHR or Ångström. OpenEmbedded is supported by the [Yocto project](https://www.yoctoproject.org/) from the Linux Foundation and provides a rock-solid basis for AsteroidOS.
+
+OpenEmbedded uses the [`bitbake`](https://docs.yoctoproject.org/bitbake/) tool which is developed in python to generate images (like asteroid-image) from packages (like asteroid-launcher-dev) built from recipes (ex: asteroid-launcher_git.bb) provided in layers (ex: meta-asteroid).
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:10]: For extra readability, it would be good to place all these technical references, like image names, package names, recipe names and layer names in backticks.

File: pages/wiki/openembedded.md [hunk 21]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+
+Layers are git repositories containing a bunch of related recipes, example of layers include board support packages (BSPs) like [meta-dory-hybris](https://github.com/AsteroidOS/meta-smartwatch/tree/master/meta-dory) or meta-rockchip which describe a way to support a new machine, but also UI layers like meta-xfce or meta-gnome which describe the building process of graphic components.
+
+A recipe describes how to fetch, patch, configure, compile, install, package (in .rpm, .deb or .ipk) and test a piece of software. Bitbake handles all those operations and dependencies between them and between recipes. In the end it can generate images or an SDK for multiple targets.
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:14]: all those operations and +the dependencies

File: pages/wiki/openembedded.md [hunk 22]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+
+Contributing to the AsteroidOS’s OE architecture mostly consists in maintaining the [meta-asteroid](https://github.com/AsteroidOS/meta-asteroid) and [meta-smartwatch](https://github.com/AsteroidOS/meta-smartwatch) repositories which contain the recipes related to AsteroidOS and the watches' BSPs.
+In depth info can be found in the [Yocto Project's manual](https://docs.yoctoproject.org/index.html)
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:17]: I think you need an extra empty line before this so that it stays a separate paragraph no ?

File: pages/wiki/openembedded.md [hunk 23]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+
 <img src="{{assets}}/img/openembedded.png" style="width:100%"/>
+
+
+# A Quick Tutorial
+
+---
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:24]: Uh that's a little awkward, ideally the CSS of titles should provide it so we don't have to hack it in each page but that's outside of the scope of this PR, feel free to ignore

File: pages/wiki/openembedded.md [hunk 24]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+
+A recipe describes how to fetch, patch, configure, compile, install, package (in .rpm, .deb or .ipk) and test a piece of software. Bitbake handles all those operations and dependencies between them and between recipes. In the end it can generate images or an SDK for multiple targets.
+
+Contributing to the AsteroidOS’s OE architecture mostly consists in maintaining the [meta-asteroid](https://github.com/AsteroidOS/meta-asteroid) and [meta-smartwatch](https://github.com/AsteroidOS/meta-smartwatch) repositories which contain the recipes related to AsteroidOS and the watches' BSPs.
+In depth info can be found in the [Yocto Project's manual](https://docs.yoctoproject.org/index.html)
+
 <img src="{{assets}}/img/openembedded.png" style="width:100%"/>
+
+
+# A Quick Tutorial
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:22]: Quick Tutorials ?

File: pages/wiki/openembedded.md [hunk 25]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+The rest of this description assumes that you have already read and followed the [Building AsteroidOS]({{rel 'wiki/building-asteroidos'}}) instructions, and that you have built AsteroidOS for at least one platform (a particular watch or the emulator), and that you are familiar with [`git`](https://www.atlassian.com/git/tutorials) and the [Linux](https://www.linux.org/forums/linux-beginner-tutorials.123/) command line.
+
+This tutorial will introduce you to the following topics:
+
+ - [Changing an existing application:](#changing-an-existing-application) You don't like how something behaves or want to fix an issue in an existing application? This will introduce you to the tooling used to allow you to make these changes.
+ - [Adding a new application:](#adding-a-new-application) Use the tooling to create a new application recipe and test it on a watch.
+ - [Sharing your changes:](#sharing-your-changes) You've made changes on your local build of AsteroidOS. How do you share these changes? This will be explained in this part of the tutorial.
+
+
+# Changing an Existing Application
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:35]: These are logically "under" "Quick Tutorials" so the titles should be smaller so the hierarchy stays consistent

File: pages/wiki/openembedded.md [hunk 26]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+devtool modify asteroid-flashlight
+```
+
+The last two lines of output will be something like these:
+
+```
+INFO: Source tree extracted to /home/developer/asteroid/build/workspace/sources/asteroid-flashlight
+INFO: Recipe asteroid-flashlight now set up to build from /home/developer/asteroid/build/workspace/sources/asteroid-flashlight
+
+```
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:105]: For the formatting to be correct, shouldn't the three backticks be one line higher ? (and then the empty line after the three backticks)

File: pages/wiki/openembedded.md [hunk 27]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+devtool undeploy-target asteroid-flashlight root@192.168.2.15
+```
+
+# Adding a New Application
+
+---
+
+Let's say you have decided to create a new application. As described in [Creating an Asteroid app]({{rel 'wiki/creating-an-asteroid-app'}}) page, the [Asteroid-helloworld](https://github.com/AsteroidOS/asteroid-helloworld) which can be used as the basis for your own new application can be cloned and then modified. So let's say you've done that and now you'd like to build the app, deploy it and test it. You can, of course, follow the instructions listed on that page, but here we'll show how to do so by working within Yocto and bitbake.
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:134]: "So let's say you've done that and now you'd like to build the app, deploy it and test it. You can, of course, follow the instructions listed on that page, but here we'll show how to do so by working within Yocto and bitbake."

I would reword that to:
"So let's say you've done that and now you'd like to integrate the app, to the Yocto and bitbake build system"

The intent here shouldn't be to provide two different ways to do the same thing just for the sake of it but rather a way to get changes upstream :)

File: pages/wiki/openembedded.md [hunk 28]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+$ devtool add hello ~/tools/AsteroidOS/asteroid-helloworld/
+NOTE: Starting bitbake server...
+NOTE: Starting bitbake server...
+INFO: Please add the following line for 'LICENSE' to a 'lib/recipetool/licenses.csv' and replace `Unknown` with the license:
+84dcc94da3adb52b53ae4fa38fe49e5d,Unknown
+NOTE: Starting bitbake server...
+INFO: Recipe /home/developer/asteroid/build/workspace/recipes/hello/hello_git.bb has been automatically created; further editing may be required to make it fully functional
+```
+
+So the tool has created a recipe based on just the name we have given and the directory containing the source. Since it seems unsure about the license and suggests that further editing might be required, let's take it's advice and edit the recipe. The simplest way to do that is again to use `devtool`:
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:152]: its*

File: pages/wiki/openembedded.md [hunk 29]
```diff
@@ -3,10 +3,243 @@ title: OpenEmbedded
...
+
+```
+devtool finish asteroid-flashlight meta-asteroid
+```
+
+This creates one patch for each new git commit and automatically updates the recipe file in the named layer (in this case, `meta-asteroid`). (Note that this is just an example and would not really be appropriate for `asteroid-flashlight` because this application is unique to AsteroidOS.)  To see what was done, you can navigate to `asteroid/src/meta-asteroid` and type `git diff`. You will see that the patch file has been added to the recipe for `asteroid-flashlight`. Also note that to ensure you don't lose any work, `devtool` will not delete the directory tree in `workspace` unless you specifically ask it to do so.
+
+## Create a pull request
+
+For an AsteroidOS application, it may be more appropriate to create a [*pull request*](https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests). To do that, we need to create a fork of the software we want to modify, create a branch, make the changes, commit and push the changes, and then create a pull request (PR). This guide doesn't describe all of these steps in detail, but describes how to use `devtool` and `git` to create a PR.
```
Comment (FlorentRevest) [inline pages/wiki/openembedded.md:207]: Maybe worth linking to something like https://opensource.com/article/19/7/create-pull-request-github ? Instead of saying "Create a pull request" "We won't tell you how to send a pull request" ?

**Review by FlorentRevest — COMMENTED** [review]
Very cool! Thanks a lot for this, this will help a lot of folks for sure

File: pages/wiki/openembedded.md (hunk 20 above)
Comment (beroset) [reply pages/wiki/openembedded.md:10]: done

File: pages/wiki/openembedded.md (hunk 21 above)
Comment (beroset) [reply pages/wiki/openembedded.md:14]: fixed

File: pages/wiki/openembedded.md (hunk 22 above)
Comment (beroset) [reply pages/wiki/openembedded.md:17]: Yes, that makes it better.

File: pages/wiki/openembedded.md (hunk 23 above)
Comment (beroset) [reply pages/wiki/openembedded.md:24]: Since you mention it, it's really not necessary to add the `---` here, and it was inconsistently applied anyway, so I've removed it from all locations on this page.

File: pages/wiki/openembedded.md (hunk 24 above)
Comment (beroset) [reply pages/wiki/openembedded.md:22]: It was intended to set the reader's expectation that it's not a book-length treatise on how to use it.

File: pages/wiki/openembedded.md (hunk 25 above)
Comment (beroset) [reply pages/wiki/openembedded.md:35]: Fixed.

File: pages/wiki/openembedded.md (hunk 26 above)
Comment (beroset) [reply pages/wiki/openembedded.md:105]: It's not strictly necessary, but it does make the output a bit neater, so I've made that change.

File: pages/wiki/openembedded.md (hunk 27 above)
Comment (beroset) [reply pages/wiki/openembedded.md:134]: better!   done.

File: pages/wiki/openembedded.md (hunk 29 above)
Comment (beroset) [reply pages/wiki/openembedded.md:207]: That's a much better link.  Thanks!

File: pages/wiki/openembedded.md (hunk 28 above)
Comment (beroset) [reply pages/wiki/openembedded.md:152]: Good spot!  fixed.

**Comment by beroset** [comment]
Added one more `#` to each topic under "Sharing your changes" to make it consistent.

File: pages/wiki/bitbake.md (hunk 1 above)
Comment (FlorentRevest) [reply pages/wiki/bitbake.md]: I agree these priorities look a bit odd, maybe something to look into :) although if it hasn't caused problems so far it can probably wait a bit

**Comment by FlorentRevest** [comment]
Thanks a lot! This will help a lot of future contributors :)
